### PR TITLE
storage: Direct mounting and unmounting

### PR DESCRIPTION
**Pull request by mvollmer**
This directly spawns "mount" and "umount" instead of using the UDisks2
methods.  This allows us to select which fstab entry to mount, and
allows us to unmount targets that are not mentioned in fstab.

Right now this matters only in corner cases, but it will become
important for Btrfs.

I have split this out to clean up the current slight "abuse" of the NFS mount monitor, and to start thinking about the corner cases that we had to ignore earlier.

- [ ] Split the process stuff out from the NFS mount monitor
- [ ] Use mount_at and unmount_at also for the "inconsistent mount" fixups, and maybe do a better job there.

### CODE REVIEW

**Comment by mvollmer** [comment]
One improvement here is that Cockpit can now tear down all mount points of a device, instead of only the first one mentioned in fstab.

**Comment by mvollmer** [comment]
The TestStorageResize.testResizeNtfs failure is https://github.com/cockpit-project/bots/issues/2392 but the error message has changed.

File: pkg/storaged/fsys-tab.jsx [hunk 1]
```diff
@@ -223,10 +223,8 @@ export function mounting_dialog(client, block, mode, forced_options) {
         }
 
         function maybe_unmount() {
-            if (mount_point_users.length > 0)
-                return client.nfs.stop_and_unmount_entry(mount_point_users, { fields: [null, old_dir] });
-            else if (block_fsys && block_fsys.MountPoints.length > 0)
-                return block_fsys.Unmount({ });
+            if (block_fsys && block_fsys.MountPoints.length > 0)
```
Comment (mvollmer) [inline pkg/storaged/fsys-tab.jsx:226]: This should check whether old_dir is in MountPoints.

Comment (mvollmer) [reply pkg/storaged/fsys-tab.jsx]: Done.

**Comment by martinpitt** [comment]
@mvollmer : Seems this slipped through the cracks as there was no review(er). Are you happy with this? If so, please push a rebase. Or was this some experiment only for developing a better udisks API? Then please mark as draft.

File: pkg/storaged/client.js [hunk 2]
```diff
@@ -594,6 +597,50 @@ client.older_than = function older_than(version) {
...
+ *
+ * But note that these functions still require there to be an entry in
+ * fstab.
+ */
+
+client.mount_at = (block, target) => {
+    const entry = utils.array_find(block.Configuration,
+                                   c => c[0] == "fstab" && utils.decode_filename(c[1].dir.v) == target);
+    if (entry)
+        return cockpit.script('mkdir -p "$2" && mount "$1" "$2" -o "$3"',
```
Comment (martinpitt) [inline pkg/storaged/client.js:615]: Why the mkdir? Does udisks do that? ISTM that the mount should just fail if the directory does not exist. Aside from being "too magic", there are also at least three traps here:

 - If the path contains a broken symlink, directory will be created (and possibly fails) in a possibly unexpected location.
 - `false && true` in POSIX shell is true (i.e. does not fail). You have to use `set -e; mkdir; mount` to get what you intend.
 - this will fail on funny characters in directory names (such as `"` or `\`). Shell is always awkward for this kind of thing.

In summary: Scrap the mkdir and the shell stuff, and use cockpit.spawn() for this.

**Review by martinpitt — CHANGES_REQUESTED** [review]
The shell issues need fixing.

A large portion of this diff is about moving existing functionality from nfs-mounts.py to mount-users.py. Could you split this out into a separate commit? Then it's easier to revert and go back to udisks API once that gets improved.

**Comment by mvollmer** [comment]
> @mvollmer : Seems this slipped through the cracks as there was no review(er). Are you happy with this? If so, please push a rebase. Or was this some experiment only for developing a better udisks API? Then please mark as draft.

This was blocked by #17188 for some time and then I got distracted by LVM2 RAID.  I put this to draft until I get back to btrfs.

**Comment by mvollmer** [comment]
> A large portion of this diff is about moving existing functionality from nfs-mounts.py to mount-users.py. Could you split this out into a separate commit?

I made it a separate PR, even, since current `main` already "abuses" the NFS code for mount users.

File: pkg/storaged/client.js (hunk 2 above)
Comment (mvollmer) [reply pkg/storaged/client.js:641]: > Why the mkdir? Does udisks do that? 

Yes.

Comment (mvollmer) [reply pkg/storaged/client.js:641]: > `false && true` in POSIX shell is true (i.e. does not fail).

This is not true (hah!) as far as I can tell.  If 'mkdir' fails, 'mount' will not be executed and cockpit.spawn will reject with an error. `bash -c "false && true"; echo $?` => `1`

> this will fail on funny characters in directory names (such as `"` or `\`)

I don't think so, since the `"$1"` construct is evaluated by bash, it is not done via cockpit.format or similar.  This works as expected and creates a directory named `f"o"o`  `bash -c 'mkdir "$1"' -- 'f"o"o'`

File: pkg/storaged/client.js [hunk 3]
```diff
@@ -602,6 +605,53 @@ client.older_than = function older_than(version) {
...
+ * it.
+ *
+ * But note that these functions still require there to be an entry in
+ * fstab.
+ */
+
+client.mount_at = (block, target) => {
+    const entry = utils.array_find(block.Configuration,
+                                   c => c[0] == "fstab" && utils.decode_filename(c[1].dir.v) == target);
+    if (entry)
```
Comment (cockpituous) [inline pkg/storaged/client.js:640]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-16991-20220825-111327-558828a7-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/client.js [hunk 4]
```diff
@@ -694,29 +744,16 @@ function nfs_mounts() {
...
-            JSON.stringify(entry)]);
+        return client.stop_mount_users(users).then(() => unmount_entry(entry));
     }
 
     function stop_and_remove_entry(users, entry) {
-        const [pids, units] = pids_and_units(users);
-        return spawn_nfs_mounts(["stop-and-remove",
-            JSON.stringify(units), JSON.stringify(pids),
-            JSON.stringify(entry)]);
+        return client.stop_mount_users(users).then(() => remove_entry(entry));
```
Comment (cockpituous) [inline pkg/storaged/client.js:752]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-16991-20220825-111327-558828a7-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/fsys-tab.jsx [hunk 5]
```diff
@@ -243,18 +241,19 @@ export function mounting_dialog(client, block, mode, forced_options) {
...
-                            return (block_fsys.Mount({ })
+                            const block = client.blocks[block_fsys.path];
+                            return (client.mount_at(block, new_dir)
                                     .catch(error => {
                                         // systemd might have mounted the filesystem for us after
                                         // unlocking, because fstab told it to.  Ignore any error
                                         // from mounting in that case.  This only happens when this
                                         // code runs to fix a inconsistent mount.
-                                        return (utils.is_mounted_synch(client.blocks[block_fsys.path])
+                                        return (utils.is_mounted_synch(block)
```
Comment (cockpituous) [inline pkg/storaged/fsys-tab.jsx:251]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-16991-20220825-111327-558828a7-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/fsys-tab.jsx [hunk 6]
```diff
@@ -243,18 +241,19 @@ export function mounting_dialog(client, block, mode, forced_options) {
...
                                                 .then(mounted_at => {
                                                     if (mounted_at == new_dir)
                                                         return;
                                                     return (undo()
-                                                            .then(() => block_fsys.Mount({ }))
+                                                            .then(() => client.mount_at(block, old_dir))
```
Comment (cockpituous) [inline pkg/storaged/fsys-tab.jsx:256]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-16991-20220825-111327-558828a7-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/client.js (hunk 3 above)
Comment (cockpituous) [inline pkg/storaged/client.js:640]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-16991-20220830-075612-e02376d4-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/client.js [hunk 7]
```diff
@@ -602,6 +605,53 @@ client.older_than = function older_than(version) {
...
+    const entry = utils.array_find(block.Configuration,
+                                   c => c[0] == "fstab" && utils.decode_filename(c[1].dir.v) == target);
+    if (entry)
+        return cockpit.script('mkdir -p "$2" && mount "$1" "$2" -o "$3"',
+                              [utils.decode_filename(block.Device), target, utils.decode_filename(entry[1].opts.v)],
+                              { superuser: true, err: "message" });
+    else
+        return Promise.reject(cockpit.format(_("No fstab entry for $0 and $1"),
+                                             utils.decode_filename(block.Device),
+                                             target));
```
Comment (cockpituous) [inline pkg/storaged/client.js:647]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-16991-20220830-075612-e02376d4-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/fsys-tab.jsx (hunk 5 above)
Comment (cockpituous) [inline pkg/storaged/fsys-tab.jsx:251]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-16991-20220830-075612-e02376d4-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/fsys-tab.jsx (hunk 6 above)
Comment (cockpituous) [inline pkg/storaged/fsys-tab.jsx:256]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-16991-20220830-075612-e02376d4-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/client.js (hunk 2 above)
Comment (martinpitt) [reply pkg/storaged/client.js:641]: Ah right: `false && true` will exit with 1, but only because it's the last exit code, not because the statement as a whole fails; `false && true; echo hello` will both run the echo *and* exit 0. I.e. these are a time bomb, as soon as someone attaches another command to that shell script it will explode. I maintain that `&&` is unintuitive and misleading, and `set -e` is more explicit and predictable.

TIL about the shell auto-magic quoting behaviour, so I give that to you. I still *hate* involving a shell in this, and I'd much rather have two chained cockpit.spawn() commands without a shell, but I won't veto it then (after moving to `set -e`, that one I insist on, sorry).

Comment (mvollmer) [reply pkg/storaged/client.js:641]: Alright!

Comment (martinpitt) [reply pkg/storaged/client.js]: Thanks! WFM now, resolving. Let's land the other PR first.

File: pkg/storaged/client.js [hunk 8]
```diff
@@ -621,6 +623,35 @@ client.stop_mount_users = (users) => {
...
+ * it.
+ *
+ * But note that these functions still require there to be an entry in
+ * fstab.
+ */
+
+client.mount_at = (block, target) => {
+    const entry = utils.array_find(block.Configuration,
+                                   c => c[0] == "fstab" && utils.decode_filename(c[1].dir.v) == target);
+    if (entry)
```
Comment (cockpituous) [inline pkg/storaged/client.js:640]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-16991-20220830-101236-2e5e4dfa-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/client.js [hunk 9]
```diff
@@ -621,6 +623,35 @@
...
+        return cockpit.script('set -e; mkdir -p "$2"; mount "$1" "$2" -o "$3"',
+                              [utils.decode_filename(block.Device), target, utils.decode_filename(entry[1].opts.v)],
+                              { superuser: true, err: "message" });
+    else
+        return Promise.reject(cockpit.format(_("No fstab entry for $0 and $1"),
+                                             utils.decode_filename(block.Device),
+                                             target));
```
Comment (cockpituous) [inline pkg/storaged/client.js:647]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-16991-20220830-101236-2e5e4dfa-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

Comment (mvollmer) [reply pkg/storaged/client.js:647]: Reaching this is more a programming error than a user error, so it's hard to test.

**Comment by martinpitt** [comment]
```
# test/verify/check-storage-mounting:427:43: F541 f-string is missing placeholders
# test/verify/check-storage-mounting:437:43: F541 f-string is missing placeholders
# test/verify/check-storage-mounting:438:44: F541 f-string is missing placeholders
```

File: pkg/storaged/client.js (hunk 9 above)
Comment (martinpitt) [reply pkg/storaged/client.js:647]: In that case, I'm in favor of dropping the gettext() -- no need to waste translator efforts on something users should never see?

File: pkg/storaged/client.js [hunk 10]
```diff
@@ -621,6 +623,35 @@ client.stop_mount_users = (users) => {
...
 };
 
+/* Direct mounting and unmounting
+ *
+ * We don't use UDisks2 for most of our mounting and unmounting in
+ * order to get better control over which entry from fstab is
+ * selected.  Once UDisks2 allows that control, we can switch back to
+ * it.
+ *
+ * But note that these functions still require there to be an entry in
```
Comment (martinpitt) [inline pkg/storaged/client.js:633]: "require an fstab entry..."

File: test/verify/check-storage-mounting [hunk 11]
```diff
@@ -407,6 +407,36 @@ ExecStart=/usr/bin/sleep infinity
...
+        self.content_row_action(1, "Mount")
+        self.dialog_wait_open()
+        self.dialog_set_val("mount_options.extra", "hurr")
+        self.dialog_apply()
+        self.dialog_wait_alert("bad option")
+        self.dialog_cancel()
+
+        # No changes should have been done to fstab
+        self.assertIn("noauto", m.execute(f"findmnt -s -n -o OPTIONS /run/foo"))
+        self.assertNotIn("hurr", m.execute(f"findmnt -s -n -o OPTIONS /run/foo"))
```
Comment (martinpitt) [inline test/verify/check-storage-mounting:438]: findmnt checks the runtime status, not fstab. I think checking findmnt is fine, but should it compare fstab before/after in addition?

**Review by martinpitt — CHANGES_REQUESTED** [review]
Danke!

File: test/verify/check-storage-mounting (hunk 11 above)
Comment (mvollmer) [reply test/verify/check-storage-mounting:438]: `findmnt -s` checks fstab.  But thanks for calling this out, I did in fact blindly copy/paste this and you could have well been right.

Comment (martinpitt) [reply test/verify/check-storage-mounting:438]: Ooh thanks, TIL -- mind using `--fstab` then? Long options are so much more readable.

Comment (mvollmer) [reply test/verify/check-storage-mounting:438]: Yes, very good point.

File: pkg/storaged/client.js (hunk 9 above)
Comment (mvollmer) [reply pkg/storaged/client.js:647]: Yes, I'll also add "Internal error:".

File: pkg/storaged/client.js (hunk 10 above)
Comment (mvollmer) [reply pkg/storaged/client.js:633]: Done, thanks.

File: pkg/storaged/client.js [hunk 12]
```diff
@@ -621,6 +621,34 @@ client.stop_mount_users = (users) => {
...
+ * selected.  Once UDisks2 allows that control, we can switch back to
+ * it.
+ *
+ * But note that these functions still require an fstab entry.
+ */
+
+client.mount_at = (block, target) => {
+    const entry = utils.array_find(block.Configuration,
+                                   c => c[0] == "fstab" && utils.decode_filename(c[1].dir.v) == target);
+    if (entry)
```
Comment (cockpituous) [inline pkg/storaged/client.js:637]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-16991-20220831-070640-0685112a-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/client.js [hunk 13]
```diff
@@ -621,6 +621,34 @@ client.stop_mount_users = (users) => {
...
+        return cockpit.script('set -e; mkdir -p "$2"; mount "$1" "$2" -o "$3"',
+                              [utils.decode_filename(block.Device), target, utils.decode_filename(entry[1].opts.v)],
+                              { superuser: true, err: "message" });
+    else
+        return Promise.reject(cockpit.format("Internal error: No fstab entry for $0 and $1",
+                                             utils.decode_filename(block.Device),
+                                             target));
```
Comment (cockpituous) [inline pkg/storaged/client.js:644]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-16991-20220831-070640-0685112a-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by martinpitt — APPROVED** [review]
Nice, thank you!

**Comment by martinpitt** [comment]
Let's please land this after tomorrow's release. That's going to RHEL and so far has no "intrusive" changes; I feel like that's a kind of  change that has the potential to break, and I don't want to do that after dev freeze.
